cmd/handler-utils.go

```diff
@@ -150,6 +150,56 @@ func extractMetadataFromHeader(ctx context.Context, header http.Header) (map[str
 	return metadata, nil
 }
 
+// extractMetadata extracts metadata from HTTP header and HTTP queryString.
+func extractMetadata(ctx context.Context, r *http.Request) (metadata map[string]string, err error) {
+	query := r.URL.Query()
+	header := r.Header
+	metadata = make(map[string]string)
+	// Extract all query values.
+	metadata, err = extractMetadataFromMap(ctx, query, metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract all header values.
+	metadata, err = extractMetadataFromMap(ctx, header, metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// Success.
+	return metadata, nil
+}
+
+// extractMetadata extracts metadata from map values.
+func extractMetadataFromMap(ctx context.Context, v map[string][]string, m map[string]string) (map[string]string, error) {
+	if v == nil {
+		logger.LogIf(ctx, errInvalidArgument)
+		return nil, errInvalidArgument
+	}
+	// Save all supported headers.
+	for _, supportedHeader := range supportedHeaders {
+		if value, ok := v[http.CanonicalHeaderKey(supportedHeader)]; ok {
+			m[supportedHeader] = value[0]
+		} else if value, ok := v[supportedHeader]; ok {
+			m[supportedHeader] = value[0]
+		}
+	}
+	for key := range v {
+		for _, prefix := range userMetadataKeyPrefixes {
+			if !strings.HasPrefix(strings.ToLower(key), strings.ToLower(prefix)) {
+				continue
+			}
+			value, ok := v[key]
+			if ok {
+				m[key] = value[0]
+				break
+			}
+		}
+	}
+	return m, nil
```

Comment:
What is the indented behaviour? return new map with values or updating existing map `m`?

Reply:
@balamurugana  So initially i wrote a single function say `extractMetadataFromQueryString` to extract the meta data from query string , then harsha suggested to have a common funtion say `extractMetadata` which reads from both query and header.

Reply:
`extractMetadataFromMap` updates existing map

Reply:
`map` is a pointer.  when caller function updates, it's key/value are changed i.e. no need to return updated map

Reply:
We should make a copy @Praveenrajmani if we wish to do that, otherwise @balamurugana suggestion is better.